### PR TITLE
nxp: kinetis: Move GPIO pin mux setting into gpio_mcux_configure

### PR DESCRIPTION
**Pull request by galak**
Move setting of PCR[MUX] for GPIO into gpio_mcux_configure() so we don't need to call pinmux_pin_set().

### CODE REVIEW

**Comment by galak** [comment]
I'll make the same change to vega once we are good with this.

**Comment by henrikbrixandersen** [comment]
I see the idea here, but I'm afraid we are introducing two different ways to set the pinmux by letting the GPIO driver do it on its own? How well will this work along side the soon-to-be pinctrl groups-based API?

**Review by henrikbrixandersen — COMMENTED** [review]
I see the idea here, but I'm afraid we are introducing two different ways to set the pinmux by letting the GPIO driver do it on its own? How well will this work along side the soon-to-be pinctrl groups-based API?

**Comment by galak** [comment]
> I see the idea here, but I'm afraid we are introducing two different ways to set the pinmux by letting the GPIO driver do it on its own? How well will this work along side the soon-to-be pinctrl groups-based API?

So some gpio drivers are effectively already doing the same thing via calls to pinmux_pin_set()

```
drivers/gpio/gpio_esp32.c:      pinmux_pin_set(data->pinmux, io_pin, PIN_FUNC_GPIO);
drivers/gpio/gpio_lpc11u6x.c:   ret = pinmux_pin_set(data->pinmux_dev, pin, func);
```

I think it makes sense to just have the GPIO pin_configure function handle this instead of trying to have every user of GPIO setting this.

**Comment by galak** [comment]
* Looking at Atmel SAM0 same thing is happening (PMUX setting)
* Atmel SAM & SAM4L seem to do this slightly indirectly (in that they have a register that says GPIO control or not)
* Silabs gecko ends up going via GPIO_PinModeSet() for both GPIO and "pin" setting.

**Comment by henrikbrixandersen** [comment]
> * Looking at Atmel SAM0 same thing is happening (PMUX setting)
> * Atmel SAM & SAM4L seem to do this slightly indirectly (in that they have a register that says GPIO control or not)
> * Silabs gecko ends up going via GPIO_PinModeSet() for both GPIO and "pin" setting.

While that is a fair reason, I still think we may run into trouble (even on the above SoCs) when the pinctrl groups API is introduced. Allow me to elaborate my concern a bit:

Once the pinctrl groups API is introduced this will be used for configuring a group of pins for a particular, pre-defined function (e.g. SPI). Let's say I have a need to, at runtime, reconfigure the SPI pins as GPIOs to perform some bit banging, e.g. to a non-standard compliant SPI device. At firmware boot, I would use the pinctrl API to set up the pins for SPI and the GPIO API to set up the GPIO direction etc. But with the above change, calling `gpio_pin_configure()` would have the side-effect of overwriting the pinmux configuration just set through the pinctrl API?

**Review by mnkp — APPROVED** [review]

**Comment by mnkp** [comment]
> Once the pinctrl groups API is introduced this will be used for configuring a group of pins for a particular, pre-defined function (e.g. SPI). Let's say I have a need to, at runtime, reconfigure the SPI pins as GPIOs to perform some bit banging, e.g. to a non-standard compliant SPI device. At firmware boot, I would use the pinctrl API to set up the pins for SPI and the GPIO API to set up the GPIO direction etc. But with the above change, calling `gpio_pin_configure()` would have the side-effect of overwriting the pinmux configuration just set through the pinctrl API?

We want to model the pin controller behavior on how it's done in Linux. In case of Linux the pins used/controlled by the driver, e.g. UART or SPI, are set up automatically during device probing. This means the driver will 'just work'. User doesn't have to worry about calling additionally pin controller driver, everything happens automatically.

GPIO is a bit different. If not for GPIO hogging (automatic pin configuration during GPIO driver initialization), which we don't support, no pins are configured when the GPIO driver is initialized. However, to follow the same approach / have similar user experience the pin controller is called when pin direction is configured during runtime.

In Linux GPIO API is designed a bit differently, every aspect of the pin configuration is controlled by a different function. So, instead of single `gpio_pin_configure()` that takes flags they have plethora of different API functions. The Linux function to set pin direction to input is [gpiod_direction_input](https://www.kernel.org/doc/html/latest/driver-api/gpio/index.html#c.gpiod_direction_input). If necessary, it will call pin controller driver to ensure the pin is configured as input and ready to work.

To sum up, having `gpio_pin_configure()` call internally pin controller to connect the pins to GPIO module follows the Linux implementation and is a user friendly approach.

Users are of course free to call pin controller API (which we don't have at the moment) during runtime to connect pins to different modules, e.g. switch pin between SPI and GPIO driver.

**Comment by henrikbrixandersen** [comment]
What I am worried about is initialisation order of GPIOs vs. other pin functions.
A couple of more use-cases:

An I2C controller, which needs to implement the `i2c_recover_bus()` API using GPIO bitbanging. I guess that would need (at least) two different `pinctrl-*` DTS properties (one for I2C SCL/SDA and one for GPIOs) and switch between those at runtime. How would the initialisation order be here? From the above, it would matter whether it configured the GPIO and _then_ the pinctrl in `init()` or if it did it the other way around.

A GPIO controller, which allows reading the GPIOs as inputs - or even using them as interrupt - even though the pinctrl is configured for another function. This is a common, but ofter overlooked, feature. The "other function" driver would take care of configuring the pinctrl for the "other function" in it's init routine, but as soon as application code configures the GPIO pin for input, it has the side-effect of re-configuring the pinctrl for GPIO, rendering the "other function" driver useless.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Blocking until we have discussed this further.

**Comment by galak** [comment]
> An I2C controller, which needs to implement the `i2c_recover_bus()` API using GPIO bitbanging. I guess that would need (at least) two different `pinctrl-*` DTS properties (one for I2C SCL/SDA and one for GPIOs) and switch between those at runtime. How would the initialisation order be here? From the above, it would matter whether it configured the GPIO and _then_ the pinctrl in `init()` or if it did it the other way around.
> 
> A GPIO controller, which allows reading the GPIOs as inputs - or even using them as interrupt - even though the pinctrl is configured for another function. This is a common, but ofter overlooked, feature. The "other function" driver would take care of configuring the pinctrl for the "other function" in it's init routine, but as soon as application code configures the GPIO pin for input, it has the side-effect of re-configuring the pinctrl for GPIO, rendering the "other function" driver useless.

for both of these cases, I wonder if we should be calling `gpio_pin_configure` at all.

For the `i2c_recover_bus` case, I expect we'd have pinctrl-0 and pinctrl-1 for the device and `i2c_recover_bus` would do something like:

```
i2c_recover_bus() {
   /* configure pins as GPIOs */
   pinset_group(dev, pinctrl-1);
   ...
   /* restore pins back to normal SCL/SDA */
   pinset_group(dev, pinctrl-0);
}
```

**Comment by henrikbrixandersen** [comment]
> for both of these cases, I wonder if we should be calling `gpio_pin_configure` at all.

We would need to set e.g. pin direction, right?

**Comment by galak** [comment]
> A GPIO controller, which allows reading the GPIOs as inputs - or even using them as interrupt - even though the pinctrl is configured for another function. This is a common, but ofter overlooked, feature. The "other function" driver would take care of configuring the pinctrl for the "other function" in it's init routine, but as soon as application code configures the GPIO pin for input, it has the side-effect of re-configuring the pinctrl for GPIO, rendering the "other function" driver useless.

Can you expand this use case with some real world examples?

**Review by henrikbrixandersen — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
